unit_tests/cpo.cc

```diff
@@ -0,0 +1,841 @@
+  a_cosine_matrices_grains_ref[1][1][0][0] = 36;
+  a_cosine_matrices_grains_ref[1][1][0][1] = 37;
+  a_cosine_matrices_grains_ref[1][1][0][2] = 38;
+  a_cosine_matrices_grains_ref[1][1][1][0] = 39;
+  a_cosine_matrices_grains_ref[1][1][1][1] = 40;
+  a_cosine_matrices_grains_ref[1][1][1][2] = 41;
+  a_cosine_matrices_grains_ref[1][1][2][0] = 42;
+  a_cosine_matrices_grains_ref[1][1][2][1] = 43;
+  a_cosine_matrices_grains_ref[1][1][2][2] = 44;
+  a_cosine_matrices_grains_ref[1][2][0][0] = 45;
+  a_cosine_matrices_grains_ref[1][2][0][1] = 46;
+  a_cosine_matrices_grains_ref[1][2][0][2] = 47;
+  a_cosine_matrices_grains_ref[1][2][1][0] = 48;
+  a_cosine_matrices_grains_ref[1][2][1][1] = 49;
+  a_cosine_matrices_grains_ref[1][2][1][2] = 50;
+  a_cosine_matrices_grains_ref[1][2][2][0] = 51;
+  a_cosine_matrices_grains_ref[1][2][2][1] = 52;
+  a_cosine_matrices_grains_ref[1][2][2][2] = 53;
+
+  volume_fractions_grains_ref[1][0] = 0.4;
+  volume_fractions_grains_ref[1][1] = 0.5;
+  volume_fractions_grains_ref[1][2] = 0.6;
+
+  volume_fractions_grains_derivatives_ref[0][0] = 0.7;
+  volume_fractions_grains_derivatives_ref[0][1] = 0.8;
+  volume_fractions_grains_derivatives_ref[0][2] = 0.9;
+
+  a_cosine_matrices_grains_derivatives_ref[0][0][0][0] = 54;
+  a_cosine_matrices_grains_derivatives_ref[0][0][0][1] = 55;
+  a_cosine_matrices_grains_derivatives_ref[0][0][0][2] = 56;
+  a_cosine_matrices_grains_derivatives_ref[0][0][1][0] = 57;
+  a_cosine_matrices_grains_derivatives_ref[0][0][1][1] = 58;
+  a_cosine_matrices_grains_derivatives_ref[0][0][1][2] = 59;
+  a_cosine_matrices_grains_derivatives_ref[0][0][2][0] = 60;
+  a_cosine_matrices_grains_derivatives_ref[0][0][2][1] = 61;
+  a_cosine_matrices_grains_derivatives_ref[0][0][2][2] = 62;
+  a_cosine_matrices_grains_derivatives_ref[0][1][0][0] = 63;
+  a_cosine_matrices_grains_derivatives_ref[0][1][0][1] = 64;
+  a_cosine_matrices_grains_derivatives_ref[0][1][0][2] = 65;
+  a_cosine_matrices_grains_derivatives_ref[0][1][1][0] = 66;
+  a_cosine_matrices_grains_derivatives_ref[0][1][1][1] = 67;
+  a_cosine_matrices_grains_derivatives_ref[0][1][1][2] = 68;
+  a_cosine_matrices_grains_derivatives_ref[0][1][2][0] = 69;
+  a_cosine_matrices_grains_derivatives_ref[0][1][2][1] = 70;
+  a_cosine_matrices_grains_derivatives_ref[0][1][2][2] = 71;
+  a_cosine_matrices_grains_derivatives_ref[0][2][0][0] = 72;
+  a_cosine_matrices_grains_derivatives_ref[0][2][0][1] = 73;
+  a_cosine_matrices_grains_derivatives_ref[0][2][0][2] = 74;
+  a_cosine_matrices_grains_derivatives_ref[0][2][1][0] = 75;
+  a_cosine_matrices_grains_derivatives_ref[0][2][1][1] = 76;
+  a_cosine_matrices_grains_derivatives_ref[0][2][1][2] = 77;
+  a_cosine_matrices_grains_derivatives_ref[0][2][2][0] = 78;
+  a_cosine_matrices_grains_derivatives_ref[0][2][2][1] = 79;
+  a_cosine_matrices_grains_derivatives_ref[0][2][2][2] = 80;
+
+  volume_fractions_grains_derivatives_ref[1][0] = 1.0;
+  volume_fractions_grains_derivatives_ref[1][1] = 1.1;
```

Comment:
Yes, it looks (and is) odd but this is just a synthetic test to see if the values are stored and loaded correctly. The first assert you suggest would be good, but I think it should be placed where the computation is actually done and not in these store and load functions. 
I have thought about the second assert (sum volume_fractions ~1) and currently I think it has value to allow a sum larger than 1. The reason for that is that the D-Rex handles all the mineral types completely separately and the only connection is the `volume fraction olivine` and the `volume fraction enstatite = 1-volume fraction olivine`. Not checking for the sum therefore allows to in one run compute for example a set of olivine grains deforming while being 100%,70% and 50% of the total grains. To be honest, it is not something I am planning on using myself, and I am fine with limiting it to the sum if there are strong opinions about it.

Reply:
Do I understand you correctly that the D-Rex-calculated CPO of X is dependent on the fraction of X but not on any of the coexisting phases. Isn't that inconsistent behaviour? What would happen if you had two types of olivine that differed only in colour (for example)? Splitting them 0:100 and 50:50 should have no effect on the resulting change in CPO. 

I'm probably misunderstanding something...

Reply:
In each computation step it only influences the derivative of the volume fraction. So final derivative grain x = volume fraction mineral 0 * derivative grain x. So it does not matter what the other mineral grains are doing, but this indirectly influences the cpo development. What the current setup allows is to use that property of the algorithm to simultaneously compute for example how the olivine cpo develops if you have 60%, 70% or 100% olivine in the mix (and you could also compute the corresponding 40%, 70% and 0% enstatite if you want. This does currently get tricky if you are interested in the elastic tensor resulting from this matrix (which is not in this pull request), because that currently assumes that every counts up to 1. 

Like I said before, I am not planning on using that, but I could see that it might be useful to others and I didn't see any harm, so that is why I didn't disallow it. This of course stops working if cpo algorithms where to be implemented in which the minerals directly interact which each other.